source/mesh_deformation/free_surface.cc

```diff
@@ -0,0 +1,417 @@
+                  continue;
+
+                fscell->get_dof_indices (cell_dof_indices);
+                fs_fe_face_values.reinit (fscell, face_no);
+                fe_face_values.reinit (cell, face_no);
+                fe_face_values[this->introspection().extractors.velocities].get_function_values(this->get_solution(), velocity_values);
+
+                cell_vector = 0;
+                cell_matrix = 0;
+                for (unsigned int point=0; point<n_face_q_points; ++point)
+                  {
+                    // Select the direction onto which to project the velocity solution
+                    Tensor<1,dim> direction;
+                    if ( advection_direction == SurfaceAdvection::normal ) // project onto normal vector
+                      direction = fs_fe_face_values.normal_vector(point);
+                    else if ( advection_direction == SurfaceAdvection::vertical ) // project onto local gravity
+                      direction = this->get_gravity_model().gravity_vector(fs_fe_face_values.quadrature_point(point));
+                    else
+                      AssertThrow(false, ExcInternalError());
+
+                    direction *= ( direction.norm() > 0.0 ? 1./direction.norm() : 0.0 );
+
+                    for (unsigned int i=0; i<dofs_per_cell; ++i)
+                      {
+                        for (unsigned int j=0; j<dofs_per_cell; ++j)
+                          {
+                            cell_matrix(i,j) += (fs_fe_face_values[extract_vel].value(j,point) *
+                                                 fs_fe_face_values[extract_vel].value(i,point) ) *
+                                                fs_fe_face_values.JxW(point);
+                          }
+
+                        cell_vector(i) += (fs_fe_face_values[extract_vel].value(i,point) * direction)
+                                          * (velocity_values[point] * direction)
+                                          * fs_fe_face_values.JxW(point);
+                      }
+                  }
+
+                mass_matrix_constraints.distribute_local_to_global (cell_matrix, cell_vector,
+                                                                    cell_dof_indices, mass_matrix, rhs, false);
+              }
+
+      rhs.compress (VectorOperation::add);
+      mass_matrix.compress(VectorOperation::add);
+
+      // Jacobi seems to be fine here.  Other preconditioners (ILU, IC) run into troubles
+      // because the matrix is mostly empty, since we don't touch internal vertices.
+      LinearAlgebra::PreconditionJacobi preconditioner_mass;
+      preconditioner_mass.initialize(mass_matrix);
+
+      SolverControl solver_control(5*rhs.size(), this->get_parameters().linear_stokes_solver_tolerance*rhs.l2_norm());
+      SolverCG<LinearAlgebra::Vector> cg(solver_control);
+      cg.solve (mass_matrix, dist_solution, rhs, preconditioner_mass);
```

Comment:
This is pre-existing, but it seems a bit fishy that we are solving a linear system that really only exists on the boundary. All rows of the matrix corresponding to interior nodes are zero. It is a bit of a miracle that the solver even succeeds, and that the Jacobi preconditioner does not error out when it takes the inverse of a zero diagonal entry.

I suspect that this could be improved somehow.